core/src/program.rs

```diff
@@ -17,16 +16,41 @@ use super::Kernel;
 
 #[derive(Clone, Debug)]
 pub struct Program {
-    mast_forest: MastForest,
+    mast_forest: Arc<MastForest>,
+    /// The "entrypoint", when set, is the root of the entire forest, i.e. a path exists from this
+    /// node to all other roots in the forest. This corresponds to the executable entry point.
+    /// Whether or not the entrypoint is set distinguishes a MAST which is executable, versus a
+    /// MAST which represents a library.
+    entrypoint: MastNodeId,
+    kernel: Arc<Kernel>,
 }
 
 /// Constructors
 impl Program {
-    pub fn new(mast_forest: MastForest) -> Result<Self, ProgramError> {
-        if mast_forest.entrypoint().is_some() {
-            Ok(Self { mast_forest })
-        } else {
-            Err(ProgramError::NoEntrypoint)
+    /// Construct a new [`Program`] from the given MAST forest and entrypoint. The kernel is assumed
+    /// to be empty.
+    pub fn new(mast_forest: Arc<MastForest>, entrypoint: MastNodeId) -> Self {
+        debug_assert!(mast_forest.get_node_by_id(entrypoint).is_some());
```

Comment:
Since I seem to recall it being explicitly documented to "do not mix up your forests and your node ids", an assertion seems appropriate to me, but I'd be fine either way.

Reply:
I agree with @bitwalker.

Related: https://github.com/0xPolygonMiden/miden-vm/pull/1359#discussion_r1653273886

Reply:
My thinking is that since this is a part of the public interface, we should provide a bit more safety as it can be used outside of the assembler. For example, we frequently construct `Program` structs directly during tests, and there could be use cases when the same happens outside of test context.

Agreed that we can't guarantee that the struct is constructed correctly, but we can catch clearly invalid constructions. I do still think returning an error would be better, and a regular `assert!` could be fine as well - e.g.:
```rust
assert!(new_root_id.0 < self.nodes.len());
```
And then we should also document the panic condition.

Reply:
I documented what I believe would be the best solution in #1368, which would guarantee that `MastNodeId`s are used properly at compile-time.

Switched to `assert!()` in the meantime.